cmd/kubectl-gadget/deploy.go

```diff
@@ -362,36 +203,129 @@ func runDeploy(cmd *cobra.Command, args []string) error {
+		return err
+	}
+	mapper := restmapper.NewDeferredDiscoveryRESTMapper(memory.NewMemCacheClient(discoveryClient))
+
+	dynamicClient, err := dynamic.NewForConfig(config)
+	if err != nil {
+		return fmt.Errorf("failed to create dynamic client: %w", err)
+	}
+
+	for _, object := range objects {
+		if daemonSet, ok := object.(*appsv1.DaemonSet); ok {
+			daemonSet.Spec.Template.Annotations["inspektor-gadget.kinvolk.io/option-hook-mode"] = hookMode
+
+			gadgetContainer := &daemonSet.Spec.Template.Spec.Containers[0]
+
+			gadgetContainer.Image = image
+
+			policy, err := stringToPullPolicy(imagePullPolicy)
+			if err != nil {
+				return err
+			}
+			gadgetContainer.ImagePullPolicy = policy
+
+			if !livenessProbe {
+				gadgetContainer.LivenessProbe = nil
+			} else {
+				gadgetContainer.LivenessProbe.InitialDelaySeconds = livenessProbeInitialDelay
+			}
+
+			for i := range gadgetContainer.Env {
+				switch gadgetContainer.Env[i].Name {
+				case "GADGET_IMAGE":
+					gadgetContainer.Env[i].Value = image
+				case "INSPEKTOR_GADGET_VERSION":
+					gadgetContainer.Env[i].Value = version
+				case "INSPEKTOR_GADGET_OPTION_HOOK_MODE":
+					gadgetContainer.Env[i].Value = hookMode
+				case "INSPEKTOR_GADGET_OPTION_FALLBACK_POD_INFORMER":
+					gadgetContainer.Env[i].Value = strconv.FormatBool(fallbackPodInformer)
+				}
+			}
+		}
+
+		if printOnly {
+			bytes, err := yaml.Marshal(object)
+			if err != nil {
+				return fmt.Errorf("problem while marshaling object: %w", err)
+			}
+			fmt.Printf("%s---\n", bytes)
+		} else {
+			err := createResource(dynamicClient, mapper, object)
+			if err != nil {
+				return fmt.Errorf("problem while creating resource: %w", err)
+			}
 		}
-		defer f.Close()
-		w = f
 	}
 
-	fmt.Fprintf(w, "%s\n---\n", resources.TracesCustomResource)
-	err = t.Execute(w, p)
+	if printOnly || noWait {
+		return nil
+	}
+
+	fmt.Println("Waiting for gadget pod(s) to be ready...")
+	k8sClient, err := k8sutil.NewClientsetFromConfigFlags(utils.KubernetesConfigFlags)
 	if err != nil {
-		return fmt.Errorf("failed to generate deploy template: %w", err)
+		return utils.WrapInErrSetupK8sClient(err)
 	}
 
-	return nil
+	// The below code (particularly how to use UntilWithSync) is highly
+	// inspired from kubectl wait source code:
+	// https://github.com/kubernetes/kubectl/blob/b5fe0f6e9c65ea95a2118746b7e04822255d76c2/pkg/cmd/wait/wait.go#L364
+	daemonSetInterface := k8sClient.AppsV1().DaemonSets(gadgetNamespace)
+	lw := &cache.ListWatch{
+		ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
+			options.LabelSelector = "k8s-app=gadget"
+
+			return daemonSetInterface.List(context.TODO(), options)
+		},
+		WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+			options.LabelSelector = "k8s-app=gadget"
+
+			return daemonSetInterface.Watch(context.TODO(), options)
+		},
+	}
+
+	ctx, cancel := watchtools.ContextWithOptionalTimeout(context.TODO(), time.Duration(timeout)*time.Second)
+	defer cancel()
```

Comment:
It's using the timeout from the undeploy command that is 30 seconds, I think it's too short for creation. Maybe 1 or 2 minutes work better?